src/index/VocabularyGenerator.cpp

```diff
@@ -0,0 +1,139 @@
+      mergeBuf[i].emplace_back();
+      mergeBuf[i].back().resize(len);
+      infiles[i].read(&(mergeBuf[i].back()[0]), len);
+      endOfFile[i] = false;
+    
+    }
+  }
+
+
+  using pair_T = std::pair<std::string, size_t>;
+  std::priority_queue<pair_T, std::vector<pair_T>, PairCompare> queue;
+  for (size_t i = 0; i < numFiles; i++) {
+    if (!mergeBuf[i].empty()) {
+      // the empty is another sanity check in case one of the partial
+      // vocabularies is empty (this should probably not happen and will be
+      // checked during next refactoring. But the performance overhead in this
+      // case is small
+      queue.push(std::make_pair(mergeBuf[i].front(), i));
+    }
+    iterators.push_back(mergeBuf[i].begin());
+  }
+
+  std::string lastWritten = "";
+  size_t totalWritten = 0;
+
+  while (!queue.empty()) {
+    auto top = queue.top();
+    queue.pop();
+
+
+    // avoid duplicates
+    if (top.first != lastWritten) {
+      lastWritten = top.first;
+      if (top.first < string({EXTERNALIZED_LITERALS_PREFIX})) {
+        outfile << top.first << std::endl;
+      } else {
+        outfileExternal << top.first << std::endl;
+      }
+      infiles[top.second].write((char*)&totalWritten, sizeof(totalWritten));
+      totalWritten++;
+    } else {
+      // always write Index (also in case of duplicates)
+      // we already have increased total written, so for the duplicate
+      // we have to subtract one again
+      auto minusOne = totalWritten - 1;
+      infiles[top.second].write((char*)&minusOne, sizeof(totalWritten));
+    }
+
+    // refill with top element from current vector
+    iterators[top.second]++;
+    if (iterators[top.second] == mergeBuf[top.second].end()) {
+      // if this condition is false, we have no more words for file i, just
+      // skip
+      if (endOfFile[top.second]) { continue;}
+
+      // refill vector from file
+      mergeBuf[top.second].clear();
+      auto i = top.second;
+      std::string word;
+      endOfFile[top.second] = true;
+      uint32_t len;
+      if (infiles[i].read((char*)&len, sizeof(len))) {
+	mergeBuf[i].emplace_back();
+	mergeBuf[i].back().resize(len);
+	infiles[i].read(&(mergeBuf[i].back()[0]), len);
+	endOfFile[i] = false;
+      }
+      iterators[i] = mergeBuf[i].begin();
+      if (mergeBuf[i].begin() != mergeBuf[i].end()) {
+	queue.push(std::make_pair(mergeBuf[i].front(), i));
+      }
+    } else {
+      queue.push(std::make_pair(*(iterators[top.second]), top.second));
+    }
+  }
+}
+
+// ____________________________________________________________________________________________
+google::sparse_hash_map<string, Id> vocabMapFromPartialIndexedFile(const string& partialFile) {
+  std::ifstream file(partialFile, std::ios_base::binary);
+  AD_CHECK(file.is_open());
+  google::sparse_hash_map<string, Id> vocabMap;
+  uint32_t len;
+  while (file.read((char*)&len, sizeof(len))) {
+    std::string word;
+    word.resize(len);
```

Comment:
Instead of initializing a default string and only then resizing it probably makes more sense to create a string of the correct size filled with zeros using the size based constructor.

As for the buffering since we read sequentially and only know the size after reading the length I don't think custom buffering makes sense here. Let's just leave that to the OS. Alternatively it may also make more sense to read the whole `partialFile` into a single buffer in one go if it's not too large. That would save some syscalls.